src/poly1305.rs

```diff
@@ -16,31 +16,115 @@
 // TODO: enforce maximum input length.
 
 use {c, chacha, constant_time, error};
+use core;
+use polyfill::slice::u32_from_le_u8;
 
-pub struct SigningContext {
-    state: State,
+struct AlignedBlockState<'a> {
+    buf: [u8; BLOCK_STATE_SIZE + 7],
+    inner: &'a mut [u8; BLOCK_STATE_SIZE],
+}
+impl<'a> AlignedBlockState<'a> {
+    pub fn new(inner: &mut [u8; BLOCK_STATE_SIZE]) -> AlignedBlockState {
+        let mut st = AlignedBlockState {
+            buf: [0u8; BLOCK_STATE_SIZE + 7],
+            inner: inner,
+        };
+        AlignedBlockState::align(&mut st.buf).copy_from_slice(&st.inner[..]);
+        st
+    }
+    fn align(xs: &mut [u8; BLOCK_STATE_SIZE + 7]) -> &mut [u8; BLOCK_STATE_SIZE] {
```

Comment:
please add a `debug_assert!(BLOCK_STATE_SIZE >= 7);` here, since this code only works when that condition is true